executor/write.go

```diff
@@ -158,7 +158,21 @@ func updateRecord(ctx context.Context, sctx sessionctx.Context, h kv.Handle, old
 		unchangedRowKey := tablecodec.EncodeRowKeyWithHandle(physicalID, h)
 		txnCtx := sctx.GetSessionVars().TxnCtx
 		if txnCtx.IsPessimistic {
-			txnCtx.AddUnchangedRowKey(unchangedRowKey)
+			txnCtx.AddUnchangedLockKey(unchangedRowKey)
+			for _, idx := range t.Indices() {
+				if !idx.Meta().Unique {
+					continue
+				}
```

Comment:
I tested with MySQL 8.0.27. It seems that the non-unique keys are locked as well:

```sql
drop table if exists t;
create table t (a varchar(10), b varchar(10), unique key(a), key(b));
insert into t values ('a0', 'b0');
begin;
update t set a = 'a0' where b = 'b0'; -- insert into t values ('a1', 'b0'); <- Blocked in another session
rollback;
```

I am not sure which behavior is reasonable.

Reply:
I think we could leave the locking for the non-unique index to other places. It should be considered with changed/non-changed rows together.